### PR TITLE
feat(float): allow enabling mouse for non-focusable window

### PR DESCRIPTION
**Pull request by zeertzjq**
Problem:  Cannot allow mouse interaction for non-focusable float window.
Solution: Add a "mouse" field to float window config.

Extracted from #30542, removing the callback part.

### CODE REVIEW

File: src/nvim/grid_defs.h [hunk 1]
```diff
@@ -80,8 +80,8 @@ struct ScreenGrid {
   // whether the compositor should blend the grid with the background grid
   bool blending;
 
-  // whether the grid can be focused with mouse clicks.
-  bool focusable;
+  // whether the grid interacts with mouse events.
+  bool mouse_enabled;
```
Comment (justinmk) [inline src/nvim/grid_defs.h:84]:
```suggestion
  bool mouse;
```

File: src/nvim/api/win_config.c [hunk 2]
```diff
@@ -129,7 +129,12 @@
 ///            fractional.
 ///   - focusable: Enable focus by user actions (wincmds, mouse events).
 ///       Defaults to true. Non-focusable windows can be entered by
-///       |nvim_set_current_win()|.
+///       |nvim_set_current_win()|, or, when the `mouse` field is set to true,
+///       by mouse events.
+///   - mouse: Specify how this window affects mouse events.
```
Comment (justinmk) [inline src/nvim/api/win_config.c:134]:
```suggestion
///   - mouse: Specify how this window treats mouse events.
```

File: src/nvim/api/win_config.c [hunk 3]
```diff
@@ -129,7 +129,12 @@
...
+///       Defaults to `focusable` value.
+///       - If false, mouse events ignore this window.
+///       - If true, mouse events interact with this window normally.
```
Comment (justinmk) [inline src/nvim/api/win_config.c:137]:
```suggestion
///       - If false, window ignores mouse events.
///       - If true, mouse interacts with this window normally.
```

**Comment by justinmk** [comment]
https://github.com/neovim/neovim/pull/30542#issuecomment-2419095045 answers my initial question:

> I'm also starting to doubt whether the callback is needed at all, as after #30741 is implemented, `vim.on_key()` along with `mouse = true` can kind of already achieve what the callback does.

Somewhat of a one-way door, unless we define `mouse: Object` in this PR, so that we can extend it later?

**Comment by zeertzjq** [comment]
Dict fields are not part of the API contract.

File: src/nvim/api/win_config.c (hunk 3 above)
Comment (zeertzjq) [reply src/nvim/api/win_config.c:137]: > window ignores mouse events.

I don't think this is correct.

File: src/nvim/api/win_config.c (hunk 2 above)
Comment (zeertzjq) [reply src/nvim/api/win_config.c:134]: No

File: src/nvim/grid_defs.h (hunk 1 above)
Comment (zeertzjq) [reply src/nvim/grid_defs.h:84]: No

File: src/nvim/api/win_config.c (hunk 2 above)
Comment (justinmk) [reply src/nvim/api/win_config.c:134]: That's a bit pedantic though. Being technically correct leads to awkward documentation without any benefit AFAICT.

**Comment by zeertzjq** [comment]
Hmm, I mean, they are not a part of the signature. If a Dict field that accepts a boolean is extended to also accept a callback, it's a backward-compatible change.

**Comment by justinmk** [comment]
> If a Dict field that accepts a boolean is extended to also accept a callback, it's a backward-compatible change.

I think we should formalize that and also extend it to the function parameters/signature. They are equivalent in practice.

**Review by justinmk — APPROVED** [review]

**Comment by errael** [comment]
FYI, On `nvim_open_win()` with `mouse = true, focusable = false`; if you left click on the floating window, the floating window gets focus, can edit/select. Isn't this wrong since `focusable` is `false`? You can use `vim.fn.win_gotoid()` to focus the window.

BTW, this is is not an issue for the popup window compatibility code since it consumes/discards these keys in a `vim.on_key()` callback. (Assuming that #30741 gets a fix.)

**Comment by zeertzjq** [comment]
> FYI, On `nvim_open_win()` with `mouse = true, focusable = false`; if you left click on the floating window, the floating window gets focus, can edit/select. Isn't this wrong since `focusable` is `false`? You can use `vim.fn.win_gotoid()` to focus the window.
> 
> BTW, this is is not an issue for the popup window compatibility code since it consumes/discards these keys in a `vim.on_key()` callback. (Assuming that #30741 gets a fix.)

Hmm, I initially intended that for the callback case, not the `true` case, so maybe the callback is needed after all.

**Comment by errael** [comment]
> initially intended that for the callback case

I can verify that returning `false` from the callback prevents a click from focusing. Instead of a `boolean` could have `mouse=` take a three value enum, for example: `ignore`, `click_will_focus`, `click_will_not_focus`.

**Pull request merged**